ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java

```diff
@@ -551,31 +663,47 @@ private static TopologyTestDriver buildStreamsTopology(final Query query,
     return tests;
   }
 
+
+  private static KsqlEngine getKsqlEngine(final SchemaRegistryClient client,
+                                          final KsqlConfig ksqlConfig) {
+      final MetaStore metaStore = new MetaStoreImpl(functionRegistry);
+
+     return new KsqlEngine(
+          new FakeKafkaTopicClient(),
+          client,
+          metaStore,
+          ksqlConfig);
+  }
+
+  private static Map<String, Object> getConfigs(final Map<String, Object> additionalConfigs) {
+      final Map<String, Object> config = new HashMap<String, Object>() {{
+          put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:0");
+          put("application.id", "KSQL-TEST");
+          put("commit.interval.ms", 0);
+          put("cache.max.bytes.buffering", 0);
+          put("auto.offset.reset", "earliest");
+          put("topology.optimization", "all");
+          put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+      }};
+      if(additionalConfigs != null){
+          config.putAll(additionalConfigs);
+      }
+      return config;
+
+  }
+
   static void shouldBuildAndExecuteQuery(final Query query) {
-    final MetaStore metaStore = new MetaStoreImpl(functionRegistry);
     final SchemaRegistryClient schemaRegistryClient = new MockSchemaRegistryClient();
 
-    final Map<String, Object> config = new HashMap<String, Object>() {{
-      put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:0");
-      put("application.id", "KSQL-TEST");
-      put("commit.interval.ms", 0);
-      put("cache.max.bytes.buffering", 0);
-      put("auto.offset.reset", "earliest");
-      put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
-    }};
+    final Map<String, Object> config = getConfigs(new HashMap<>());
     final Properties streamsProperties = new Properties();
     streamsProperties.putAll(config);
     final KsqlConfig ksqlConfig = new KsqlConfig(config);
```

Comment:
KSQL saves configs alongside queries on the command topic, and uses them to produce the final config used to build the query. That way, if we make a breaking change in the ksql engine, we have a way to preserve the old behavior for stuff that's running. We should replicate this in the test by persisting the config map (as returned by ```KsqlConfig.getKsqlConfigPropsWithSecretsObfuscated```), along with each topology.

Then, before building we should run:

```
Map<String, Object> persistedConfig = <config saved w/ topology>
KsqlConfig ksqlConfig = new KsqlConfig(config).overrideBreakingConfigsWithOriginalValues(persistedConfig)
```

Once we start releasing more often, we'll probably also want to build each query with the current defaults (no overriding) and check that the topology matches whatever the last topology dumped was (with its configs overridden as described above), and then only write a new set of query files out if something has changed.